numpy/core/src/umath/ufunc_object.c

```diff
@@ -4475,14 +4476,60 @@ _get_fixed_signature(PyUFuncObject *ufunc,
  * need to "cast" to string first).
  */
 static int
-resolve_descriptors(int nop,
+resolve_descriptors(int nop, int nin,
```

Comment:
Is it useful to pass in `nin`? I though that one could just get that with `ufunc->nin`? (Similarly, is `nop` necessary?)

Reply:
More generally, I wonder if it wouldn't make more sense to just pass through everything one could possibly want to the `_raw` version. If a ufunc wants raw, perhaps it should deal with the nitty-gritty (including even `check_safety`, or set that up differently so one is not stuck with `original_dtypes`). In that case, the name is more appropriate too...

The question then is if there is anything missing from the current argument list above that could possibly be useful? Logically, I guess it is just the outputs -- which could be easily attached by passing through `full_args` rather than just `full_args.in`.

If so, I'd suggest to remove `nin` and add `ufunc_full_args full_args`

Reply:
Yes we don't have to pass `nin` and `nout`, it is on the ufunc and `ufuncimpl`.  I guess for a single `nop` it seemed easier, with both it is getting unclear.

Right, I would be happy to add outputs in principle.  Right now I did this because I limited to true scalars and those are simply never outputs (i.e. the `i < nout` is only a sanity check).
One reason was that it isn't necessary trivial to distinguish a scalar from a 0-D object and the *original* 0-D object may need to be converted to an array to do anything with it so at that point I would have to pass both if you were to actually use that information reasonably.  One thing that could work is checking for whether it is a scalar based on the DType, even when not abstract, and otherwise pass the array.)

I havne't thought of integrating the cast-safety check.  Right now we have a "disable cast safety" special path for logical ufuncs loop (casting to bool is always OK).
I would be happy with passing the cast-safety and either allowing or forcing you to take care of it.  I.e. we can remove the flag or if you want to deal with it you use set the flag so NumPy doesn't check again.
(I am not aware of where it would be useful to side-step the check except for that logical `bool, bool -> bool` loop.)

I would agree that there might also be a good reason to allow a "let me do everything" path where we let the user do as much as reasonable (e.g. also the full iteration).  I think we should have that although finding the right API for it might not be trivial (unless we launch back into/through effectively python?!).
It isn't super easy to get feature parity, but would allow a lot of weirder stuff and speed experimentation that might be cool.  (Also just making allowing more things to look like a ufunc and use the protocol.)

Reply:
Yes, it is tricky. And indeed, if one wants more control, it would probably also be necessary to know not just the ufunc but also whether it is a call, `at`, `reduce`, etc. At which point it is no longer obvious this is the right place. Still, my tendency would be to add `full_args` instead of the inputs and just pass everything on, i.e., let the implementation deal with `input_scalars` (and `original_types`). But perhaps keep the check on cast safety, since that is what the function has to return anyway.

Reply:
Well, I prefer passing everything on as `PyObject *const *` (I will add the const) rather than two tuples.  Just because I think that is the more convenient API usually, and I also think changing the way we do this in the ufunc code itself would be an improvement probably (create the tuple when we need it, we don't need it often).

Passing the `method` additionally, would seem OK to me.  I am not sure it is useful/needed, but I could imagine it can be (if just for printing a nicer custom error message).

My uncertainty is how implementations should deal with weird types (non-arrays but also not obvious scalars), so I am not quite sure what is the best thing to pass on there if that is what they want.
Thus, my maybe silly inclination to punt on it, but maybe rename the argument.  We could pass all arguments as it is, we currently just keep them `NULL` when it isn't quite clear what is best.

Reply:
@ngoldbaum maybe you have an opinion?  Since you were also thinking about what one could do usefully with the original input information.

Reply:
I mostly think for the `_raw` version it makes sense to pass through whatever goes in `resolve_descriptors`, though since these are guaranteed to be proper tuples, there is no real reason not to do `PySequence_Fast_ITEMS` to get you your `PyObject **`.

Reply:
Ah, TBH, I don't care about that bit of dissonance right now between the `resolve_descriptors` internal helper and the API function, partially because I don't like `full_args` :), it's just what we currently have.

The main question remains if we should pass on everything even when the original argument for non-scalars is a very messy thing (you may still have to convert it to an array to do anything useful, OTOH, what do you want to do that this is necessary?).

In the spirit of keeping things "raw" forwarding, it probably makes sense to pass `NULL` for the `PyObject *const *` (rather than an array of `NULL`), though.

Reply:
I also don't like `full_args` much, so indeed better not to expose that! And makes sense to allow `NULL` for the whole array of arrays.

Reply:
This is already enough to implement unit exponentiation with an integer and fixed-width text multiply with an integer, right? That already solves the two major pain points I'm aware of with the current API. Sure, it technically could make sense to multiply a string array by an integer array but I also think it's reasonable for numpy to error in that case.

So, for now, I don't think it's worth it to try to work out exactly how this should work until someone has a real-world use case, and let's just do the scalar-only special case because we know we definitely need at least that for signed/unsigned integer comparison inside numpy and other cool things people want to do in user dtypes.

Reply:
@ngoldbaum - I think it is really a question of where the scalar selection should live. My sense is that if you do it as in this PR, you are limiting the API of the new function needlessly and just setting yourself up for needing a `resolve_descriptors_really_raw` in the future (while still having to support this one). I think the code here is easy enough that whatever routines need this can just quickly implement it themselves.

Reply:
I could see it either way. I think sebastian has a point that it would be awfully inconvenient to be handed an arbitrary PyObject array-like or scalar and, in the general case, need to re-implement numpy's array conversion logic. 

That said, a unit exponentiation implementation would just need to check if it was handed a python or numpy float or int type, so in practice it may not be so terrible to see the raw operands.

Reply:
The code here is not very complicated, so I think that, like you said, in practical cases, going through the inputs is not that hard; it just keeps options open.

Reply:
I think I am not sure that it doesn't close as many options as it opens, although I guess it opens options to experiment without changing NumPy code.
* We can still start passing all original arguments without breaking the API.  The slot name will just be a misnomer (but we can add another name/alias).
  * I would like to at the very least pass on all scalar objects, anyway!
* If we don't decide now, we don't have to make up our mind on how we think about it and e.g. consider passing the arrays rather than original object when it isn't the obvious scalar.  (Because to me passing an original array-like seems useless, you don't want to convert _again_!  OTOH, I don't love passing arrays, because it is nice if the API is as array object agnostic as possible).

If the danger of `resolve_descriptors_really_raw` is low, remember that introducing this and deprecating the old function isn't super scary.  Yes, its slow and annoying, but it is possible and would probably just consist of adding a "default" `resolve_descriptors_really_raw` implementation that calls `resolve_descriptors_with_scalars` (and a deprecation warning will be given at loop registration time).  The complexity is more like a slightly more tedious Python deprecation now and not an "arrg, this is C-API" desparation.

Reply:
OK, true, and I don't want to make too big a deal of it. 

But I think my suggestion, passing on the outputs as well, remains easier to explain and work with. And while you are right that passing in the pre-conversion-to-array input may be quite useless from an `ndarray` perspective, who knows, maybe it is really handy from the perspective of a more random (sub)class that overrides `__array_ufunc__` and needs some strange dtype resolution as well. (For quantities, like @ngoldbaum, we had this nice procedure of the final unit of an operation only depending on the units of the arguments, which works for all but `np.power`. For that case, the scalar value is all that is needed, but maybe for other cases one needs the class shape or so.)

Or, to turn the argument around, how does passing on all inputs and outputs hurt the current use cases?

Reply:
:shrug:  Yes.  I am fine if you wish to say that I largely don't like it because I don't want you to do things like `np.emath` (inspect values to guess result dtype).  But even after you might convince me that there are actual use-cases where this would be useful – I am sure there are –, I am confident that doing it this way will still be a bad protocol for reasons beyond being brittle for alternate array-objects and conceptually not really extending to non-arrays at all.

One extension, we could also consider is pivoting to is passing `char *const *scalar_vals` which would hold the raw (possibly unaligned?) original dtype data pointer (whatever that is) or the Python objects for the current Python scalars.
If we talk about shapes, what I think you actually want is passing the core dimensions (reduction size/shape), not original objects (you can't really infer much from there), this is even more interesting in the `get_loop()` (where we have the data available in principle).

(In a sense, a lot of this to me comes back to: Yes, we probably want a "let me do everything", or even an alternative pivot where we allow more ufunc-like objects to dispatch via `__array_ufunc__` to embrace a split of "ufunc as an implementation" and "ufunc as a protocol" a bit.)

Reply:
My problem remains that the current approach feels super directed just at the particular use case. If that was done privately, without opening API for it, that would be perfectly fine. But if one adds new API, it would seem nice to be not so directed at one particular use case.

But maybe your idea for the extension is a way to resolve this. Though would one perhaps generalize it slightly? Either the original dtype or the object if it had no dtype (i.e., `getattr(obj, "dtype", obj)`). Or maybe just anything that did not have a `dtype`, and NULL otherwise? (I'm not quite sure any more what processing has happened to the dtype before we get to this stage; maybe passing the original dtype is best if it can be different.)

Reply:
Yes, it is directed at the string multiplication/unit power use-case, because that is the biggest hole I am aware off and there is some overlap in the solution.
Maybe the main problem is that it isn't quite clear yet what the correct pattern is of enabling that use case.

I can live with getting it wrong (i.e. deprecating it later).  But we really need a solution for NEP 50's sake more than anything.  So I am OK either way:
1. Just pass all arguments and make it public.  I dislike it, so I would add a scary warning to it (maybe an underscore), but in the end we are all grown-ups here after all.
2. Just make it private for now.
3. We iterate once and maybe land on the `char **` solution (may make this use-case slightly more complex, but generalize the string mul/unit power one maybe).

---

Just FYI... Probably clear, but let me summarize what happens right now:
* `__array_ufunc__`, this is the only place where you can avoid any form of array conversion.
* Array coercion.  This is a two step process:
  * Finding shape and dtype.  Which includes calling `__array__()` when necessary.
  * If necessary, creating a new array and filling it in.
  Especially for scalars, we don't have to do the second step in principle.  But for most array-like objects, we currently don't allow delaying the full conversion anyway.
* Promotion based on the DType *classes*.
  * This find the `ArrayMethod`
* The resolution step (this one).  We clearly have the "original" dtype instances available here.  (For Python integers, this right now means its could be an `int64`, `uint64` or `object` dtype).
  * *Normally*, we will convert/cast these to the actual loop DType class so that the user doesn't have to worry about that.  (This is normally a de-facto no-op, because most dtypes are non-parametric)

---

> My problem remains that the current approach feels super directed just at the particular use case.

Yes, it clearly is directly at solving exactly two things.  First, side-stepping the preparation because in its current form it fails for abstract DTypes and, second, trying to create/prepare a path forward for unit power and string multiply.  (maybe it isn't particularly good at that, because you may want to support 3-4 things: `unit**4`, `unit**np.uint8(4)` and `unit**np.array([4])`; where the last IMO should include array-likes if you allow it).

Reply:
I can't think of a clear solution to this...  So, I made the slot raise an error when used as public API (effectively making it private).